include/onnxruntime/core/common/span_utils.h

```diff
@@ -64,7 +67,22 @@ constexpr auto AsSpan(const T (&arr)[N]) {
   return details::AsSpanImpl(arr, N);
 }
 
-template<class T>
+template <class T>
 inline gsl::span<const T> EmptySpan() { return gsl::span<const T>(); }
 
+template <class U, class T>
+[[nodiscard]] inline gsl::span<U> ReinterpretAsSpan(gsl::span<T> src) {
+  // adapted from gsl-lite span::as_span():
+  // https://github.com/gsl-lite/gsl-lite/blob/4720a2980a30da085b4ddb4a0ea2a71af7351a48/include/gsl/gsl-lite.hpp#L4102-L4108
+  Expects(src.size_bytes() % sizeof(U) == 0);
+  return gsl::span<U>(reinterpret_cast<U*>(src.data()), src.size_bytes() / sizeof(U));
 }
+
+template <class T1, size_t Extent1, class T2, size_t Extent2>
+[[nodiscard]] inline bool SpanEq(gsl::span<T1, Extent1> a, gsl::span<T2, Extent2> b) {
```

Comment:
note that MS GSL's span has an additional `size_t Extent` template parameter. std::span does too.